### ISSUE TITLE
quarkus: corrupts h2 database

### ISSUE THREAD

**Issue by nimo23** [quarkusio/quarkus#5865]
**Describe the bug**
I use h2 database provider:
```
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-hibernate-orm</artifactId>
    </dependency>
    <dependency>
        <groupId>io.quarkus</groupId>
        <artifactId>quarkus-jdbc-h2</artifactId>
    </dependency>
```

I get the following error:

```
2019-11-29 00:41:40,899 ERROR [io.qua.dev.DevModeMain] (main) Failed to start Quarkus: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:222)
	at io.quarkus.runtime.Application.start(Application.java:94)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:143)
	at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:177)
	at io.quarkus.dev.DevModeMain.start(DevModeMain.java:95)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:66)
Caused by: javax.persistence.PersistenceException: [PersistenceUnit: default] Unable to build Hibernate SessionFactory
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.persistenceException(FastBootEntityManagerFactoryBuilder.java:113)
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:67)
	at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:54)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
	at javax.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
	at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:109)
	at io.quarkus.hibernate.orm.runtime.JPAConfig.startAll(JPAConfig.java:57)
	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.startAllPersistenceUnits(HibernateOrmRecorder.java:82)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits31.deploy_0(HibernateOrmProcessor$startPersistenceUnits31.zig:51)
	at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits31.deploy(HibernateOrmProcessor$startPersistenceUnits31.zig:70)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:150)
	... 5 more
Caused by: org.hibernate.exception.GenericJDBCException: Unable to open JDBC Connection for DDL execution
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:47)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.resource.transaction.backend.jta.internal.DdlTransactionIsolatorJtaImpl.<init>(DdlTransactionIsolatorJtaImpl.java:62)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorBuilderImpl.buildDdlTransactionIsolator(JtaTransactionCoordinatorBuilderImpl.java:46)
	at org.hibernate.tool.schema.internal.HibernateSchemaManagementTool.getDdlTransactionIsolator(HibernateSchemaManagementTool.java:175)
	at org.hibernate.tool.schema.internal.HibernateSchemaManagementTool.buildGenerationTargets(HibernateSchemaManagementTool.java:135)
	at org.hibernate.tool.schema.internal.SchemaDropperImpl.doDrop(SchemaDropperImpl.java:110)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:145)
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:73)
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:320)
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:462)
	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:65)
	... 14 more
Caused by: org.h2.jdbc.JdbcSQLException: Row {1} not found in primary index "PUBLIC.SYS_DATA: 1227" [90143-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
	at org.h2.message.DbException.get(DbException.java:179)
	at org.h2.message.DbException.get(DbException.java:155)
	at org.h2.mvstore.db.MVPrimaryIndex.getRow(MVPrimaryIndex.java:202)
	at org.h2.mvstore.db.MVTable.getRow(MVTable.java:492)
	at org.h2.mvstore.db.MVSecondaryIndex$MVStoreCursor.get(MVSecondaryIndex.java:493)
	at org.h2.engine.Database.open(Database.java:768)
	at org.h2.engine.Database.openDatabase(Database.java:286)
	at org.h2.engine.Database.<init>(Database.java:280)
	at org.h2.engine.Engine.openSession(Engine.java:66)
	at org.h2.engine.Engine.openSession(Engine.java:179)
	at org.h2.engine.Engine.createSessionAndValidate(Engine.java:157)
	at org.h2.engine.Engine.createSession(Engine.java:140)
	at org.h2.engine.Engine.createSession(Engine.java:28)
	at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:351)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:124)
	at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:103)
	at org.h2.Driver.connect(Driver.java:69)
	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:200)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:390)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:372)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:65)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:835)
```

Even when using `drop-and-create`, quarkus throws this error. Connecting to my database with _h2 admin console_ also fails, it shows the following error (which differs from the stacktrace above):

`General error: "java.lang.IllegalStateException: Unable to read the page at position 5283703127288069 [1.4.200/6]" [50000-200] HY000/50000`

**The only solution is:** start the **jee wildfly server** which connects to the database (with the same database connection settings) and it magically solves the issues above. That means: I stopped the wildfly server, then logged into my h2-database and I can start quarkus successfully. However sometimes quarkus throws this exception above again and I cannot log in to my database anymore, only if I repeat this step by starting wildfly server which magically (recovers?) the database instance. After starting, I can stop widlfly. Then starting quarkus with h2-connection works. Maybe it has to do with the (better?) driver which wildfly uses internally: 

```
// from wildfly standalone.xml
<driver name="h2" module="com.h2database.h2">
    <xa-datasource-class>org.h2.jdbcx.JdbcDataSource</xa-datasource-class>
</driver>
```

There seems to be an issue with `quarkus-jdbc-h2`.

**Expected behavior**
Connect to the database without corrupting it. Connect to the database by _h2-admin-console_ while quarkus is running or throwing exceptions.

**To Reproduce**
I dont know, the error throws sporadically but often..

**Configuration**
```properties
quarkus.datasource.url = jdbc:h2:~/h2/testdb;AUTO_SERVER=TRUE
quarkus.datasource.driver = org.h2.Driver
quarkus.datasource.username = test
quarkus.datasource.password = test
quarkus.hibernate-orm.database.generation = drop-and-create
quarkus.hibernate-orm.dialect = org.hibernate.dialect.H2Dialect
```

**Environment (please complete the following information):**
 - Output of `java -version`: jdk 12
 - Quarkus version or git rev: 1.0 FINAL

**Comment by Sanne** [comment]
I suspect the option `AUTO_SERVER=TRUE` could be problematic.

Quarkus would initialize a connection pool, and it's quite fast to boot this up; this implies there's possible multiple connections attempting to create a server - backed by filesystem locks - all at approximately the same time.

This is just a theory though, I'd need to find some time to be able to reproduce this.

You might want to try the helpers we created to start an H2 in a more "standard" way:
 - https://quarkus.io/guides/datasource#in-memory-databases

**Comment by Sanne** [comment]
BTW, Quarkus doesn't patch the H2 database so if it's being corrupted, it's either a bug in H2 or a consequence of "driving" it into an unsupported way. So while I'd like to investigate, it might also make sense to report this to the H2 issue tracker.

**Comment by nimo23** [comment]
I used the same h2 database connection with `AUTO_SERVER=TRUE` in wildfly since a few years without any problems.

Anyway, I reported it also to h2 https://github.com/h2database/h2database/issues/2310

**Comment by Sanne** [comment]
I'll close this: the problem is caused by that property, people should not use it in this way.

**Comment by nimo23** [comment]
>the problem is caused by that property, people should not use it in this way.

@Sanne Should not use it in this way **only in quarkus or in general**? `AUTO_SERVER=TRUE` is a valid configuration option in h2 and needed when **Automatic Mixed Mode** is needed. Then please add this limitation to quarkus documentation so users are aware of that they cannot use all the config options of h2.

**Comment by Sanne** [comment]
you shouldn't use it in this way **in general**. The only thing about Quarkus is that it's so fast to boot that it triggers the problem with more likelyhood, but you might still have it on other platforms too, if you have bad timing luck e.g. on a fast machine with many cores.

**Comment by nimo23** [comment]
>you shouldn't use it in this way in general.

Interesting, I ve never faced such errors in the past for a few years in wildfly so I had good timing luck or slow machines:)

And which connection string should I use when I want to use **automatic mixed mode**. All the valid URL connections are listed in http://www.h2database.com/html/features.html under section **"Database URL Overview"**. 

I use the "Automatic mixed mode"-URL:

```
// Automatic mix mode syntax
jdbc:h2:<url>;AUTO_SERVER=TRUE
jdbc:h2:~/test;AUTO_SERVER=TRUE 
```

**H2 officially supports:**
- Embedded mode
- Server-mode
- Mixed mode

And my url is correctly fomed:

`jdbc:h2:~/h2/testdb;AUTO_SERVER=TRUE`


**So how can I use the H2 "Mixed mode" in quarkus?**

**Comment by Sanne** [comment]
When you start a new process, this is generally run "in parallel" with your main code. When you do start anything, you should wait for it to be in a sane state before using it, such as "finished booting".

I don't know what's the best way to wait for H2 to have finished booting; you might want to check if it logged something like "ready", as other databases would.

In general the problem with "AUTO_SERVER" is that you don't know if it's going to start something or not as you've not specificed if it should start - or not :) You're intentionally leaving that to H2 to decide.

N.B. MySQL and MariaDB have similar issues: it's actually not easy to test them even using Docker as they will take a good time to start, and don't make it easy for developers to know if it's done starting yet. If you try connecting before these databases actually finished starting, you'll get encoding exceptions.

It's a general difficulty, but the only thing that's making this more interesting for Quarkus is that we're too fast to start, so this issue gets more noticeable - but it wasn't strictly correct to do this on WilldFly either.

**Comment by nimo23** [comment]
Ok, good to know. 

> If you try connecting before these databases actually finished starting, **you'll get encoding exceptions.**

And these exceptions will stop quarkus to proceed instead of waiting for connection and trying to reconnect. 


Would it be a good idea to have at least a property in which I can quarkus instruct **to wait with a time out.** For example, 

```
// waits at most 200 ms for connection, after timeout the exception is thrown
quarkus.db.connection.timeout=200
```

It would work something like this 

`phaser.awaitAdvanceInterruptibly(0, 200, TimeUnit.MILLISECONDS);`


This would also solve issues you described for MySql or MariaDb and all other databases in the future **where the _db connection process_ is slower than _quarkus bootup_**. What do you think? Good idea?

Should I create a new issue like "add db.connection.timeout"?

**Comment by nimo23** [comment]
Btw, we also have "**MicroProfile Fault Tolerance"** with **Timeout** and **Retry**..maybe quarkus can make use of it when connecting to the database..and we can control this timeout and retry by quarkus properties:

```
quarkus.db.connection.timout=200
quarkus.db.connection.retry=3
```

**Comment by Sanne** [comment]
We already have connection timeouts, but this is manifesting as a corrupted answer; it wouldn't be wise to swallow those and try again. It needs to either be fixed by the databases (both client and server could have some improvements here).

**Comment by nimo23** [comment]
>It needs to either be fixed by the databases (both client and server could have some improvements here).

Got it, thanks. So https://github.com/h2database/h2database/issues/2310 is the right place:)
